Map reference scale

Rachael-E opened the pull request:
@tschie this is for the map reference scale sample: please can you take you look? Thanks!

Rachael-E committed:
Create MapReferenceScaleSample.java

Rachael-E committed:
Set up MVC structure for sample

Rachael-E committed:
Set up basic map with MVC structure

Rachael-E committed:
Added label which updates scale of map as it changes

Rachael-E committed:
Set up means to change the map view to the selected reference scale

Rachael-E committed:
Set up check boxes for map layers

Rachael-E committed:
Update MapReferenceScaleController.java

Rachael-E committed:
Update MapReferenceScaleController.java

Rachael-E committed:
Allow string in combobox, split to return double to feed into map reference scale

Rachael-E committed:
Link checkbox with map layer to set reference scale

Rachael-E committed:
Method descriptions and remove unused imports

Rachael-E committed:
Fix indents

Rachael-E committed:
Create MapReferenceScale.png

Rachael-E committed:
Create README.md

Rachael-E committed:
Convert markdown to HTML

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+    scaleComboBox.getSelectionModel().select(2);
+
+    // create a label to display current scale of the map
+    mapView.addMapScaleChangedListener(event -> {
```
can be simplified to
```java
mapView.addMapScaleChangedListener(event -> 
    scaleLabel.setText("Current Map Scale 1:" + Math.round(mapView.getMapScale()))
);
```

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+   * Sets the scale symbols (i.e. if the layer will honor the reference scale or not)
+   * for each layer within the map
+   */
+  private void setScaleSymbolsOnSelectedLayers() {
```
This method should only be called once. You could just move the logic into the `addDoneLoadingListener` lambda above.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+    // set the map to the map view
+    mapView.setMap(map);
+    map.setReferenceScale(250000);
+    scaleLabel.setText("LOADING...");
```
You could alternatively hide the vboxes at the start and show them when the map is done loading. I would also put a spinner at the center of the map to indicate it is loading. With "loading" labels, it's unclear what is loading. With a spinner over the map, it's clearer.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+    loadingLabel.setText("LOADING...");
+
+    // set up the combobox to have 1:50k, 100k, 250k and 500k reference scales
+    scaleComboBox.getItems().addAll("1:50000", "1:100000", "1:250000", "1:500000");
```
You can do this in the FXML:
```xml
<ComboBox fx:id="scaleComboBox" maxWidth="Infinity" value="1:250000">
   <items>
        <FXCollections fx:factory="observableArrayList">
             <String fx:value="1:50000"/>
             <String fx:value="1:100000"/>
             <String fx:value="1:250000"/>
             <String fx:value="1:500000"/>
        </FXCollections>
   </items>
</ComboBox>
```

tschie replied:
Seem my suggestion below if you want to use `Double`s

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+      CheckBox checkBox = new CheckBox(layer.getName());
+      checkBox.setSelected(true);
+      // add the checkboxes to the VBox
+      vBox.getChildren().add(checkBox);
```
This is fine for this sample. For reference though, you could also use a list view with a check box list cell type like in this sample: https://github.com/Esri/arcgis-runtime-samples-java/blob/master/src/main/java/com/esri/samples/featurelayers/statistical_query_group_and_sort/StatisticalQueryGroupAndSortController.java#L98.

That would come in handy if you ever wanted to add/remove layers or if you had a lot of layers and needed to scroll.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+      // add the checkboxes to the VBox
+      vBox.getChildren().add(checkBox);
+
+      FeatureLayer featureLayer = (FeatureLayer) layer;
```
Should probably check if `layer instanceof FeatureLayer` before creating a check box.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+   * Takes the reference scale from the combobox, and sets it as the map's reference scale.
+   */
+  @FXML
+  private void handleScaleButtonClicked() {
```
I don't seem to be able to select a new reference scale without also setting the map scale. Seems like they should be independent.

I would think there would be two handlers: one when the combo box selected item changes to set the reference scale, and another when the button is clicked to set the map scale to the reference scale.

Is that not how the other SDKs are doing it?

Rachael-E replied:
Good catch, I'd missed that from the earlier design: just checked with Trevor and the behaviour should be separated. Working on it!

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,139 @@
+
+    // get string value from the combobox, convert to double
+    String selectedString = scaleComboBox.getSelectionModel().getSelectedItem();
+    String[] splitString = selectedString.split(":");
```
Instead of doing this, you could give the combo box a `Double` type and use a string converter for the display:

```java
@FXML
private ComboBox<Double> scaleComboBox;

//...

scaleComboBox.setConverter(new StringConverter<Double>() {
      @Override
      public String toString(Double value) {
        return "1:" + Math.round(value);
      }

      @Override
      public Double fromString(String string) {
        // not needed
        return null;
      }
});
```

If you used my suggestion above about the FXML, you'd need to update that too:
```xml
<ComboBox fx:id="scaleComboBox" maxWidth="Infinity">
      <value>
           <Double fx:value="250000"/>
      </value>
      <items>
            <FXCollections fx:factory="observableArrayList">
                    <Double fx:value="50000"/>
                    <Double fx:value="100000"/>
                    <Double fx:value="250000"/>
                    <Double fx:value="500000"/>
           </FXCollections>
      </items>
</ComboBox>
```

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/README.md:
```diff
@@ -0,0 +1,45 @@
+<li>FeatureLayer</li>
+</ul>
+
+<h2 id="additionalinformation">Additional Information</h2>
```
You can remove the id attribute

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/README.md:
```diff
@@ -0,0 +1,45 @@
+
+<h4>Tags</h4>
+
+<p>Maps &amp; Scenes, reference scale</p>
```
I think that symbol would be fine as `&`

tschie reviewed (approved):
Looks good! Just a few optional suggestions

Rachael-E committed:
Addressed comments from review. ComboBox now double, vBoxes will load after map has loaded

Rachael-E committed:
Add a method to handle setting the reference scale when interacting with the combobox and the scale sync button.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+  @FXML
+  private Label scaleLabel;
+  @FXML
+  private Label loadingLabel;
```
Might want to give this another name now. Also, since the text is only set once, it can be done in the FXML.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  /**
+   * Set the map reference scale when a selection is made using the combobox
```
Javadoc should have punctuation

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+   * Takes the reference scale from the combobox, and sets it as the map's reference scale.
+   */
+  @FXML
+  private void handleScaleButtonClicked() {
```
I was thinking it would be like this:

```java
   /**
   * Set the map's reference scale to the scale selected in the combo box.
   */
  @FXML
  private void handleComboBoxSelection() {
    
    map.setReferenceScale(scaleComboBox.getSelectionModel().getSelectedItem());
  }

  /**
   * Set the current viewpoint scale to the map's reference scale.
   */
  @FXML
  private void handleScaleButtonClicked() {
    
    // get the center of the current viewpoint extent
    Point centerPoint = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).getTargetGeometry().getExtent().getCenter();
    // get the map's current reference scale
    double currentReferenceScale = mapView.getMap().getReferenceScale();
    // set a viewpoint with the scale at the map's reference scale
    Viewpoint newViewPoint = new Viewpoint(centerPoint, currentReferenceScale);
    mapView.setViewpointAsync(newViewPoint);
  }
```

The first handler already sets the reference scale when a selection is made. No need to set it again in the second handler (just get it from the map).

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+    });
+
+    map.addDoneLoadingListener(() -> {
+
```
should check `map.getLoadStatus() == LoadStatus.LOADED` and show alert otherwise

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+        CheckBox checkBox = new CheckBox(layer.getName());
+        checkBox.setSelected(true);
+        // add the checkboxes to the VBox
+        layerVBox.getChildren().add(checkBox);
```
Move the check box logic inside `layer instanceof FeatureLayer`. You don't want to create check boxes for non feature layers. Basically we just ignore any non feature layers.

tschie commented on src/main/java/com/esri/samples/map/map_reference_scale/MapReferenceScaleController.java:
```diff
@@ -0,0 +1,162 @@
+
+          FeatureLayer featureLayer = (FeatureLayer) layer;
+          // set if the feature layer will honor the reference scale
+          checkBox.setOnAction(event -> {
```
any time you have a single line expression in a lambda you can omit the braces.

tschie reviewed (changes requested):
Good changes. Almost there!

Rachael-E committed:
Code updates following review

removed unneeded label, added if loaded statement, and refactored map reference scale. Also updated Readme to account for behavior change.

Rachael-E committed:
move checkbox logic in with instanceof check

tschie committed:
Update MapReferenceScaleController.java
